rfd/0195-vnet-windows.md

```diff
@@ -0,0 +1,339 @@
+The client process retains control of all Teleport client methods and user keys
+so that it can easily perform MFA prompts and hardware key signatures using
+existing code.
+
+## Details
+
+### UX
+
+The goal is to keep the VNet UX identical on Windows and MacOS.
+Teleport Connect will have support VNet with the best UX for details like MFA
+prompts and error reporting.
+`tsh vnet` will be supported on Windows just as it is on MacOS, but building
+a first-class CLI UX will not be a goal, Connect will be the recommended client.
+
+In keeping with the MacOS experience, this design avoids administrator (UAC)
+prompts each time VNet is started, only requiring a prompt the very first time
+VNet is started on a Windows client machine.
+Errors and any user prompts will be identical on each OS, with a goal of reusing
+as much code between operating systems as possible so that this happens by
+default.
+
+### Virtual Network Interface
+
+VNet requires a virtual network interface to be created on the host OS,
+the specific kind of interface we use is typically called a TUN interface.
+VNet configures IP routes on the host so that IP traffic for a set of configured
+CIDR ranges are bound to the TUN.
+VNet then reads and writes IP packets to and from the TUN interface to handle
+TCP connections to Teleport apps and UDP connections to VNet's internal DNS server.
+
+Because Windows does not natively support TUN interfaces the same way that MacOS
+and Linux do, we will leverage an open-source driver wintun.dll to provide the
+TUN interface.
+The DLL is available for download from https://www.wintun.net/ as a signed DLL.
+It has a custom license that permits commercial use as long as we use their
+signed DLL via the provided API.
+Another open-source library that we already use on MacOS provides a common Go
+interface for the TUN, this is `golang.zx2c4.com/wireguard/tun`.
+
+The signed `wintun.dll` file will be distributed with Connect and installed in
+the same directory as `tshd`.
+
+### Windows Service
+
+Creating the TUN interface requires administrator rights on Windows.
+To create the TUN without requiring the user to run Connect as administrator or
+requiring a UAC prompt each time VNet starts, we will install and run VNet as a
+Windows Service.
+This is a similar concept to the Launch Daemon that we use on MacOS.
+
+The actual exe that runs the service will be `tsh` (known as `tshd` when
+installed with Connect), it will be started with a specific argument to run the
+VNet Windows service.
+Windows Services are installed and controlled by the Service Control Manager (SCM).
+The first time VNet is started on a specific Windows machine, `tsh` will
+re-execute itself with administrator rights via a UAC prompt, and then make a
+request to the SCM to install itself as a Service.
+
+We will install the Service with a security descriptor that allows the
+installing user's SID (user ID) to launch the service without elevated privileges.
```

Comment:
I asked about this under the other PR (https://github.com/gravitational/teleport/pull/50468#discussion_r1907350467), but I guess this is a better place to do this. Does it mean that only a single user on the device can install and use the service?

Reply:
updated this in the latest to support multiple per-user installations and per-machine installations.